osutil/synctree.go

```diff
@@ -0,0 +1,190 @@
+// invocations.
+//
+// For example, if the glob "snap.$SNAP_NAME.*" is used then the
+// caller should avoid trying to populate any directories matching
+// "snap.*".
+//
+// If an error occurs, all matching files are removed from the tree.
+//
+// A list of changed and removed files is returned, as relative paths
+// to the base directory.
+func EnsureTreeState(baseDir string, globs []string, content map[string]map[string]*FileState) (changed, removed []string, err error) {
+	// Sanity check globs before doing anything
+	if _, index, err := matchAny(globs, "foo"); err != nil {
+		return nil, nil, fmt.Errorf("internal error: EnsureTreeState got invalid pattern %q: %s", globs[index], err)
+	}
+	// Sanity check directory paths and file names in content dict
+	for relPath, dirContent := range content {
+		if filepath.IsAbs(relPath) {
+			return nil, nil, fmt.Errorf("internal error: EnsureTreeState got absolute directory %q", relPath)
+		}
+		if ok, index := matchAnyComponent(globs, relPath); ok {
+			return nil, nil, fmt.Errorf("internal error: EnsureTreeState got path %q that matches glob pattern %q", relPath, globs[index])
+		}
+		for baseName := range dirContent {
+			if filepath.Base(baseName) != baseName {
+				return nil, nil, fmt.Errorf("internal error: EnsureTreeState got filename %q in %q, which has a path component", baseName, relPath)
+			}
+			if ok, _, _ := matchAny(globs, baseName); !ok {
+				return nil, nil, fmt.Errorf("internal error: EnsureTreeState got filename %q in %q, which doesn't match any glob patterns %q", baseName, relPath, globs)
+			}
+		}
+	}
+	// Find all existing subdirectories under the base dir.  Don't
+	// perform any modifications here because, as it may confuse
+	// Walk().
+	subdirs := make(map[string]bool)
+	err = filepath.Walk(baseDir, func(path string, fileInfo os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if !fileInfo.IsDir() {
+			return nil
+		}
+		relPath, err := filepath.Rel(baseDir, path)
+		if err != nil {
+			return err
+		}
+		subdirs[relPath] = true
+		return nil
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	// Ensure we process directories listed in content
+	for relPath := range content {
+		subdirs[relPath] = true
+	}
+
+	maybeEmpty := []string{}
+
+	var firstErr error
+	for relPath := range subdirs {
+		dirContent := content[relPath]
+		path := filepath.Join(baseDir, relPath)
+		if err := os.MkdirAll(path, 0755); err != nil {
+			firstErr = err
+			break
+		}
+		dirChanged, dirRemoved, err := EnsureDirStateGlobs(path, globs, dirContent)
+		changed = appendWithPrefix(changed, relPath, dirChanged)
+		removed = appendWithPrefix(removed, relPath, dirRemoved)
+		if err != nil {
+			firstErr = err
+			break
+		}
+		if len(removed) != 0 {
+			maybeEmpty = append(maybeEmpty, relPath)
+		}
+	}
+	// As with EnsureDirState, if an error occurred we want to
+	// delete all matching files.  This means reprocessing
```

Comment:
".. all matching files under the whole baseDir hierarchy. This means emptying also ..."